sentinelhub/download/models.py

```diff
@@ -0,0 +1,257 @@
+    :param use_session: A flag that specifies if the download request will require a session header
+    :param data_type: An expected file format of downloaded data. Default is `MimeType.RAW`
+    :param save_response: A flag defining if the downloaded data will be saved to disk. Default is `False`.
+    :param data_folder: A folder path where the fetched data will be (or already is) saved. Default is `None`
+    :param filename: A custom filename where the data will be saved. By default, data will be saved in a folder
+        which name are hashed request parameters.
+    :param return_data: A flag defining if the downloaded data will be returned as an output of download procedure.
+        Default is `True`.
+    :param extra_params: Any additional parameters.
+    """
+
+    url: Optional[str] = None
+    headers: JsonDict = field(default_factory=dict)
+    request_type: RequestType = RequestType.GET
+    post_values: Optional[JsonDict] = None
+    use_session: bool = False
+    data_type: MimeType = MimeType.RAW
+    save_response: bool = False
+    data_folder: Optional[str] = None
+    filename: Optional[str] = None
+    return_data: bool = True
+    extra_params: Dict[str, Any] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        """Additional parsing of init parameters."""
+        self.request_type = RequestType(self.request_type)
+        self.data_type = MimeType(self.data_type)
+
+    def __hash__(self) -> int:
+        """This dataclass is mutable, but we still assign its id as its hash."""
+        return id(self)
+
+    def raise_if_invalid(self) -> None:
+        """Method that raises an error if something is wrong with request parameters
+
+        :raises: ValueError
+        """
+        if self.save_response and self.data_folder is None:
+            raise ValueError(
+                "Data folder is not specified. Please give a data folder name in the initialization of your request."
+            )
+
+    def get_request_params(self, include_metadata: bool = False) -> JsonDict:
+        """Provides parameters that define the request in form of a dictionary
+
+        :param include_metadata: A flag defining if also metadata parameters should be included, such as headers and
+            current time
+        :return: A dictionary of parameters
+        """
+        params = {"url": self.url, "payload": self.post_values}
+        if include_metadata:
+            params = {**params, "headers": self.headers, "timestamp": dt.datetime.now().isoformat()}
+        return params
+
+    def get_hashed_name(self) -> str:
+        """It takes request url and payload and calculates a unique hashed string from them.
+
+        :return: A hashed string
+        """
+        params = self.get_request_params(include_metadata=False)
+        hashable = json.dumps(params)
+
+        return hashlib.md5(hashable.encode("utf-8")).hexdigest()
```

Comment:
This change will change all the hashes, right? Perhaps to keep in mind when writing a changelog

Reply:
Nope, the hashes will stay exactly the same.

Reply:
ah, because `extra_params` aren't included in the `get_request_params`